analyzers/src/SonarAnalyzer.CSharp/Rules/SillyMathematicalComparison.cs

```diff
@@ -0,0 +1,108 @@
+        private const string MessageFormat = "Comparison to integral constant is useless; the constant is outside the range of type '{0}'";
+
+        private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterNodeAction(
+                CheckBinary,
+                SyntaxKind.GreaterThanExpression,
+                SyntaxKind.GreaterThanOrEqualExpression,
+                SyntaxKind.LessThanExpression,
+                SyntaxKind.LessThanOrEqualExpression);
+
+        private static void CheckBinary(SonarSyntaxNodeReportingContext context)
+        {
+            var binary = (BinaryExpressionSyntax)context.Node;
+            var typeIdentifier = string.Empty;
+            if (ShouldRaise(context.SemanticModel, binary.Left, binary.Right, ref typeIdentifier))
+            {
+                context.ReportIssue(Diagnostic.Create(Rule, binary.GetLocation(), typeIdentifier));
+            }
+
+            if (ShouldRaise(context.SemanticModel, binary.Right, binary.Left, ref typeIdentifier))
+            {
+                context.ReportIssue(Diagnostic.Create(Rule, binary.GetLocation(), typeIdentifier));
+            }
+        }
+
+        private static bool ShouldRaise(SemanticModel model, SyntaxNode expectedConstant, SyntaxNode other, ref string typeIdentifier) =>
+            TryGetConstant(model, expectedConstant, out var constant)
+            && model.GetSymbolInfo(other).Symbol.GetSymbolType() is { } symbol
+            && TryGetRange(symbol, out typeIdentifier, out var min, out var max)
+            && (constant < min || constant > max);
+
+        private static bool TryGetConstant(SemanticModel model, SyntaxNode expectedConstant, out double constant)
+        {
+            if (FindConstant(model, expectedConstant, Convert.ToDouble) is { } constValue)
+            {
+                constant = constValue;
+                return true;
+            }
+
+            constant = default;
+            return false;
+        }
+
+        private static T? FindConstant<T>(SemanticModel semanticModel, SyntaxNode node, Func<object, T> converter) where T : struct
+            => semanticModel.GetSymbolInfo(node).Symbol is var symbol
+                && !IsFieldOrPropertyOutsideSystemNamespace(symbol)
+                && !IsEnum(symbol)
+                && CSharpFacade.Instance.FindConstantValue(semanticModel, node) is { } value
+                && ConversionHelper.TryConvertWith(value, converter, out T typedValue)
+                ? typedValue
+                : null;
+
+        private static bool TryGetRange(ITypeSymbol typeSymbol, out string typeIdentifier, out double min, out double max)
+        {
+            if (typeSymbol.Is(KnownType.System_Single))
```

Comment:
Only cover `float`, since `double` cannot "overflow" in comparisons and `Half` is not "comparable" with float or double ( `someHalf < someFloat` is a compiler time error).